### PR TITLE
Exposing Azure HTTP requests method in Azurecore extension for external consumption

### PR DESCRIPTION
**Pull request by aasimkhan30**

### CODE REVIEW

File: extensions/azurecore/src/azurecore.d.ts [hunk 1]
```diff
@@ -75,10 +82,17 @@ declare module 'azurecore' {
...
+		 * Makes Azure REST requests to create, retireve, update or delete access to azure service's resources.
+		 * For reference to different service URLs, See https://docs.microsoft.com/en-us/rest/api/?view=Azure
+		 * @param account The azure account used to acquire access token
+		 * @param subscription The subscription under azure account where the service will perform operations.
+		 * @param serviceUrl The endpoint for the service. See https://docs.microsoft.com/en-us/rest/api/?view=Azure
+		 * @param requestType Http request method. Currently GET, PUT, POST and DELETE methods are supported.
+		 * @param requestBody Optional request body to be used in PUT and POST requests.
+		 * @param ignoreErrors When this flag is set the method will not throw any runtime or service errors and will return the error in errors array.
+		 */
+		makeAzureRestRequest(account: azdata.Account, subscription: azureResource.AzureResourceSubscription, serviceUrl: string, requestType: HttpRequestMethod, requestBody?: any, ignoreErrors?: boolean): Promise<AzureRestResponse>;
```
Comment (Charles-Gagnon) [inline extensions/azurecore/src/azurecore.d.ts:95]: For now I think we should control the specific endpoints by having them pass in an enum from this list https://github.com/microsoft/azuredatastudio/blob/main/extensions/azurecore/src/account-provider/providerSettings.ts#L12 and then pick the actual endpoint from that based on the current configured cloud - this way the callers don't have to know about the different cloud endpoints.

Comment (Charles-Gagnon) [reply extensions/azurecore/src/azurecore.d.ts:95]: Although you'll have to figure out what to do about the region-specific host you're using currently. I can't imagine that's going to be a common scenario so I'm hesitant to build that in for everything but I suppose better to do that now and have it even if we don't use it for most things.

File: extensions/sql-migration/src/sqlmigration.d.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+export interface MigrationControllerProperties {
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/sqlmigration.d.ts:6]: If you aren't exposing this to other extensions I'd say that having a typings file like this is unnecessary.

File: extensions/azurecore/src/azureResource/utils.ts [hunk 3]
```diff
@@ -347,24 +340,28 @@ export async function makeHttpRequest(account: azdata.Account, subscription: azu
...
 			break;
-		case HttpRequestType.DELETE:
-			response = await axios.delete(url, config);
+		case HttpRequestMethod.DELETE:
+			response = await axios.delete(serviceUrl, config);
 			break;
 	}
 
-	if (response.status !== 200) {
+	if (response.status >= 400) {
```
Comment (alanrenmsft) [inline extensions/azurecore/src/azureResource/utils.ts:364]: what does this mean?

Comment (aasimkhan30) [reply extensions/azurecore/src/azureResource/utils.ts:364]: https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html

Based on this doc, I am treating any status with code >= 400 as error codes.

File: extensions/azurecore/src/azurecore.d.ts [hunk 4]
```diff
@@ -75,10 +82,18 @@ declare module 'azurecore' {
...
-		createMigrationController(account:azdata.Account, subscription: azureResource.AzureResourceSubscription, resourceGroupName: string, regionName: string, controllerName: string, ignoreErrors?:boolean): Promise<CreateMigrationControllerResult>;
-		getMigrationControllerAuthKeys(account: azdata.Account, subscription: azureResource.AzureResourceSubscription, resourceGroupName: string, regionName: string, controllerName: string, ignoreErrors?: boolean): Promise<GetMigrationControllerAuthKeysResult>;
-
+		/**
+		 * Makes Azure REST requests to create, retireve, update or delete access to azure service's resources.
+		 * For reference to different service URLs, See https://docs.microsoft.com/en-us/rest/api/?view=Azure
+		 * @param account The azure account used to acquire access token
+		 * @param subscription The subscription under azure account where the service will perform operations.
+		 * @param serviceUrl The endpoint for the service. See https://docs.microsoft.com/en-us/rest/api/?view=Azure
```
Comment (alanrenmsft) [inline extensions/azurecore/src/azurecore.d.ts:90]: first of all, the view=azure in the url doesn't work, I have to manually change the view to Azure. 
the url you get from the website is the full URI, the comment for the parameter should mention how to get the partial serviceURL, and the if overrideHost is not true, serviceUrl must start with a '/'

File: extensions/azurecore/src/azureResource/utils.ts (hunk 3 above)
Comment (Charles-Gagnon) [reply extensions/azurecore/src/azureResource/utils.ts:364]: Why are you changing this now? This doesn't seem related to your changes. And given that we're not handling any of the other cases (such as the redirects) here I'm not sure we get any benefit from changing this.

File: extensions/azurecore/src/azurecore.d.ts (hunk 4 above)
Comment (Charles-Gagnon) [inline extensions/azurecore/src/azurecore.d.ts:90]: serviceUrl doesn't really make sense - this isn't normally a URL. Something like `requestPath` would be better

Comment (Charles-Gagnon) [inline extensions/azurecore/src/azurecore.d.ts:90]: How are callers supposed to set which endpoint to use if they don't want to specify it themselves?

File: extensions/azurecore/src/azurecore.d.ts [hunk 5]
```diff
@@ -75,10 +82,18 @@ declare module 'azurecore' {
...
+		 * @param requestType Http request method. Currently GET, PUT, POST and DELETE methods are supported.
+		 * @param requestBody Optional request body to be used in PUT and POST requests.
+		 * @param ignoreErrors When this flag is set the method will not throw any runtime or service errors and will return the errors in errors array.
+		 * @param overrideHost Set this flag when you are passing host as a part of serviceUrl. Default host is 'https://management.azure.com'
```
Comment (Charles-Gagnon) [inline extensions/azurecore/src/azurecore.d.ts:94]: I'd rather we do something like have a param that's either one of the enum values or a string if the caller wants to specify the endpoint themselves. That would make it easier for most callers to use this.

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: extensions/azurecore/src/azurecore.d.ts (hunk 5 above)
Comment (alanrenmsft) [reply extensions/azurecore/src/azurecore.d.ts:94]: yeah, please review it and if it is possible to extract out the some fixed pattern from the url you are expecting , I think what Charles said makes sense.

File: extensions/azurecore/src/azureResource/utils.ts (hunk 3 above)
Comment (aasimkhan30) [reply extensions/azurecore/src/azureResource/utils.ts:366]: There is an endpoint that return 201. I am adding this change to handle that edge case

Comment (Charles-Gagnon) [reply extensions/azurecore/src/azureResource/utils.ts:366]: Then you should do >= 200 && <= 299 as those are the successful responses.

File: extensions/azurecore/src/azureResource/utils.ts [hunk 6]
```diff
@@ -288,18 +288,11 @@ export async function getSelectedSubscriptions(appContext: AppContext, account?:
...
-	DELETE
-}
-
 /**
- * Make a HTTP request to Azure REST apis.
+ * Make an HTTP request to Azure REST apis.
  */
-export async function makeHttpRequest(account: azdata.Account, subscription: azureResource.AzureResourceSubscription, ignoreErrors: boolean = false, url: string, requestType: HttpRequestType, requestBody?: any): Promise<HttpRequestResult> {
-	const result: HttpRequestResult = { response: {}, errors: [] };
+export async function makeHttpRequest(account: azdata.Account, subscription: azureResource.AzureResourceSubscription, path: string, requestType: HttpRequestMethod, requestBody?: any, ignoreErrors: boolean = false, host?: string): Promise<AzureRestResponse> {
```
Comment (Charles-Gagnon) [inline extensions/azurecore/src/azureResource/utils.ts:294]: I don't think host should be optional - there isn't really a use case for that currently and I'd rather callers know exactly what endpoint they're connecting to for now

And a couple other notes here for future things : 

1. If it was then I would suggest you use the default value in the parameter instead - that makes it much clearer what the default is
2. You should always be documenting in the comments what defaults are if there are any

File: extensions/azurecore/src/azureResource/utils.ts [hunk 7]
```diff
@@ -288,18 +288,11 @@ export async function getSelectedSubscriptions(appContext: AppContext, account?:
...
-enum HttpRequestType {
-	GET,
-	POST,
-	PUT,
-	DELETE
-}
-
 /**
- * Make a HTTP request to Azure REST apis.
+ * Make an HTTP request to Azure REST apis.
```
Comment (Charles-Gagnon) [inline extensions/azurecore/src/azureResource/utils.ts:292]: Please update this with parameters

File: extensions/azurecore/src/azureResource/utils.ts [hunk 8]
```diff
@@ -347,24 +340,30 @@ export async function makeHttpRequest(account: azdata.Account, subscription: azu
 		validateStatus: () => true // Never throw
 	};
 
-	let response;
+	let requestUrl: string;
+	if (host) {
+		requestUrl = `${host}${path}`;
```
Comment (Charles-Gagnon) [inline extensions/azurecore/src/azureResource/utils.ts:345]: I'd suggest looking into using the URI class for making the full URL - that's safer and should handle either starting with a / or not. 

https://github.com/microsoft/azuredatastudio/blob/main/src/vs/vscode.d.ts#L1306

(test it out - if it doesn't handle starting with or without / then probably not worth it though)

File: extensions/azurecore/src/azureResource/utils.ts [hunk 9]
```diff
@@ -384,51 +383,19 @@ export async function makeHttpRequest(account: azdata.Account, subscription: azu
...
 		blobContainers: response?.response?.data?.value ?? [],
 		errors: response.errors ? response.errors : []
 	};
 }
 
 export async function getFileShares(account: azdata.Account, subscription: azureResource.AzureResourceSubscription, storageAccount: azureResource.AzureGraphResource, ignoreErrors: boolean): Promise<GetFileSharesResult> {
-	const apiEndpoint = `https://management.azure.com/subscriptions/${subscription.id}/resourceGroups/${storageAccount.resourceGroup}/providers/Microsoft.Storage/storageAccounts/${storageAccount.name}/fileServices/default/shares?api-version=2019-06-01`;
-	const response = await makeHttpRequest(account, subscription, ignoreErrors, apiEndpoint, HttpRequestType.GET);
+	const apiEndpoint = `/subscriptions/${subscription.id}/resourceGroups/${storageAccount.resourceGroup}/providers/Microsoft.Storage/storageAccounts/${storageAccount.name}/fileServices/default/shares?api-version=2019-06-01`;
+	const response = await makeHttpRequest(account, subscription, apiEndpoint, HttpRequestMethod.GET, undefined, ignoreErrors);
```
Comment (Charles-Gagnon) [inline extensions/azurecore/src/azureResource/utils.ts:396]: These need to be updated to match the new signature right?

File: extensions/azurecore/src/azurecore.d.ts [hunk 10]
```diff
@@ -75,10 +82,18 @@ declare module 'azurecore' {
...
 		getFileShares(account: azdata.Account, subscription: azureResource.AzureResourceSubscription, storageAccount: azureResource.AzureGraphResource, ignoreErrors?: boolean): Promise<GetFileSharesResult>;
-		getMigrationController(account: azdata.Account, subscription: azureResource.AzureResourceSubscription, resourceGroupName: string, regionName: string, controllerName: string, ignoreErrors?: boolean): Promise<GetMigrationControllerResult>;
-		createMigrationController(account:azdata.Account, subscription: azureResource.AzureResourceSubscription, resourceGroupName: string, regionName: string, controllerName: string, ignoreErrors?:boolean): Promise<CreateMigrationControllerResult>;
-		getMigrationControllerAuthKeys(account: azdata.Account, subscription: azureResource.AzureResourceSubscription, resourceGroupName: string, regionName: string, controllerName: string, ignoreErrors?: boolean): Promise<GetMigrationControllerAuthKeysResult>;
-
+		/**
+		 * Makes Azure REST requests to create, retireve, update or delete access to azure service's resources.
```
Comment (Charles-Gagnon) [inline extensions/azurecore/src/azurecore.d.ts:86]: retrieve

File: extensions/azurecore/src/azureResource/utils.ts (hunk 3 above)
Comment (alanrenmsft) [reply extensions/azurecore/src/azureResource/utils.ts:379]: image

I think you should only treat 2xx as success, 3xx is for redirection.

**Review by alanrenmsft — APPROVED** [review]
looks good to me once you address the feedback for the status code

File: extensions/azurecore/src/azureResource/utils.ts (hunk 3 above)
Comment (aasimkhan30) [reply extensions/azurecore/src/azureResource/utils.ts]: I was not considering redirections as errors.

File: extensions/azurecore/src/azureResource/utils.ts [hunk 11]
```diff
@@ -376,7 +376,7 @@ export async function makeHttpRequest(account: azdata.Account, subscription: azu
 			break;
 	}
 
-	if (response.status >= 400) {
+	if (response.status >= 300) {
```
Comment (alanrenmsft) [inline extensions/azurecore/src/azureResource/utils.ts:379]: and <200?

File: extensions/azurecore/src/azureResource/utils.ts (hunk 3 above)
Comment (Charles-Gagnon) [reply extensions/azurecore/src/azureResource/utils.ts]: But unless we handle them then it's not going to be what the user wants - so until we do making them errors makes sense to me.

File: extensions/azurecore/src/azurecore.d.ts [hunk 12]
```diff
@@ -75,10 +82,18 @@ declare module 'azurecore' {
...
+		/**
+		 * Makes Azure REST requests to create, retrieve, update or delete access to azure service's resources.
+		 * For reference to different service URLs, See https://docs.microsoft.com/en-us/rest/api/?view=Azure
+		 * @param account The azure account used to acquire access token
+		 * @param subscription The subscription under azure account where the service will perform operations.
+		 * @param path The path for the service starting from '/subscription/..'. See https://docs.microsoft.com/en-us/rest/api/azure/.
+		 * @param requestType Http request method. Currently GET, PUT, POST and DELETE methods are supported.
+		 * @param requestBody Optional request body to be used in PUT and POST requests.
+		 * @param ignoreErrors When this flag is set the method will not throw any runtime or service errors and will return the errors in errors array.
+		 * @param host Use this to override the host. The default host is https://management.azure.com'
```
Comment (Charles-Gagnon) [inline extensions/azurecore/src/azurecore.d.ts:94]: Extra ' on the end

File: extensions/sql-migration/src/api/azure.ts [hunk 13]
```diff
@@ -82,22 +82,44 @@ export async function getBlobContainers(account: azdata.Account, subscription: S
...
-export async function getMigrationController(account: azdata.Account, subscription: Subscription, resourceGroupName: string, regionName: string, controllerName: string): Promise<azureResource.MigrationController> {
+export async function getMigrationController(account: azdata.Account, subscription: Subscription, resourceGroupName: string, regionName: string, controllerName: string): Promise<MigrationController> {
 	const api = await getAzureCoreAPI();
-	let result = await api.getMigrationController(account, subscription, resourceGroupName, regionName, controllerName, true);
-	return result.controller!;
+	const host = `https://${regionName}.management.azure.com`;
+	const path = `/subscriptions/${subscription.id}/resourceGroups/${resourceGroupName}/providers/Microsoft.DataMigration/Controllers/${controllerName}?api-version=2020-09-01-preview`;
+	const response = await api.makeAzureRestRequest(account, subscription, path, azurecore.HttpRequestMethod.GET, undefined, true, host);
+	if (response.errors.length > 0) {
+		throw response.errors;
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/api/azure.ts:91]: You should throw an error - not the raw error array.

File: extensions/sql-migration/src/api/azure.ts [hunk 14]
```diff
@@ -82,22 +82,44 @@ export async function getBlobContainers(account: azdata.Account, subscription: S
...
-	let result = await api.createMigrationController(account, subscription, resourceGroupName, regionName, controllerName, true);
-	return result.controller!;
+	const host = `https://${regionName}.management.azure.com`;
+	const path = `/subscriptions/${subscription.id}/resourceGroups/${resourceGroupName}/providers/Microsoft.DataMigration/Controllers/${controllerName}?api-version=2020-09-01-preview`;
+	const requestBody = {
+		'location': regionName
+	};
+	const response = await api.makeAzureRestRequest(account, subscription, path, azurecore.HttpRequestMethod.PUT, requestBody, true, host);
+	if (response.errors.length > 0) {
+		throw response.errors;
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/api/azure.ts:106]: Same here - throw actual Error

File: extensions/sql-migration/src/api/azure.ts [hunk 15]
```diff
@@ -82,22 +82,44 @@ export async function getBlobContainers(account: azdata.Account, subscription: S
...
-export async function getMigrationControllerAuthKeys(accounts: azdata.Account, subscription: Subscription, resourceGroupName: string, regionName: string, controllerName: string): Promise<azurecore.GetMigrationControllerAuthKeysResult> {
+export async function getMigrationControllerAuthKeys(account: azdata.Account, subscription: Subscription, resourceGroupName: string, regionName: string, controllerName: string): Promise<GetMigrationControllerAuthKeysResult> {
 	const api = await getAzureCoreAPI();
-	let result = await api.getMigrationControllerAuthKeys(accounts, subscription, resourceGroupName, regionName, controllerName, true);
-	return result;
+	const host = `https://${regionName}.management.azure.com`;
+	const path = `/subscriptions/${subscription.id}/resourceGroups/${resourceGroupName}/providers/Microsoft.DataMigration/Controllers/${controllerName}/ListAuthKeys?api-version=2020-09-01-preview`;
+	const response = await api.makeAzureRestRequest(account, subscription, path, azurecore.HttpRequestMethod.POST, undefined, true, host);
+	if (response.errors.length > 0) {
+		throw response.errors;
```
Comment (Charles-Gagnon) [inline extensions/sql-migration/src/api/azure.ts:117]: And here

**Review by Charles-Gagnon — APPROVED** [review]
Couple last things to address but otherwise looks good

File: extensions/azurecore/src/azureResource/utils.ts (hunk 11 above)
Comment (Charles-Gagnon) [reply extensions/azurecore/src/azureResource/utils.ts:379]: https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#1xx_informational_response

File: extensions/azurecore/src/azureResource/utils.ts [hunk 16]
```diff
@@ -376,7 +376,7 @@ export async function makeHttpRequest(account: azdata.Account, subscription: azu
 			break;
 	}
 
-	if (response.status >= 300) {
+	if (response.status <= 200 || response.status >= 299) {
```
Comment (Charles-Gagnon) [inline extensions/azurecore/src/azureResource/utils.ts:379]: This is saying 200 and 299 are errors which isn't correct. It should be < 200 || > 299

File: extensions/azurecore/src/azureResource/utils.ts [hunk 17]
```diff
@@ -288,18 +288,19 @@ export async function getSelectedSubscriptions(appContext: AppContext, account?:
...
-enum HttpRequestType {
-	GET,
-	POST,
-	PUT,
-	DELETE
-}
-
 /**
- * Make a HTTP request to Azure REST apis.
+ * Makes Azure REST requests to create, retireve, update or delete access to azure service's resources.
```
Comment (Charles-Gagnon) [inline extensions/azurecore/src/azureResource/utils.ts:292]: retrieve

File: extensions/azurecore/src/azureResource/utils.ts [hunk 18]
```diff
@@ -288,18 +288,19 @@ export async function getSelectedSubscriptions(appContext: AppContext, account?:
...
+ * For reference to different service URLs, See https://docs.microsoft.com/en-us/rest/api/?view=Azure
```
Comment (Charles-Gagnon) [inline extensions/azurecore/src/azureResource/utils.ts:293]: remove en-us from this and the other links in this comment

**Pull request merged**